src/graph_visualization.cpp

```diff
@@ -0,0 +1,550 @@
+                    break;
+                }
+            }
+            if (flag == 1) {
+                break;
+            }
+        }
+    }
+
+    for (int i = 0; i < deleted_edges.size(); i++) {
+        tree_node first_node = get_tree_node_by_id(nodes, deleted_edges[i].first);
+        tree_node second_node = get_tree_node_by_id(nodes, deleted_edges[i].second);
+        
+        if (first_node.layer == second_node.layer + 1) {
+            for (int j = 0; j < nodes.size(); j++) {
+                if (nodes[j].id == first_node.id) {
+                    nodes[j].succ.push_back(second_node.id);
+                }
+                if (nodes[j].id == second_node.id) {
+                    nodes[j].pred.push_back(first_node.id);
+                }
+            }
+        } else {
+            int count_dummy = first_node.layer - (second_node.layer + 1);
+            add_dummy_node(nodes, first_node, second_node, -1);
+            for (int j = 1; j < count_dummy; j++) {
+                add_dummy_node(nodes, nodes.back(), second_node, -1);
+            }
+        }
+    }
+}
+
+void coord_init(std::vector<tree_node> &nodes) {
+    for (int i = 0; i < nodes.size(); i++) {
+        nodes[i].y = (nodes[i].layer + 1) * DISTANCE;
+        nodes[i].x = (nodes[i].number + 1) * DISTANCE;
+    }
+}
+
+void transformation_graph(std::vector<tree_node> &nodes) {
+    std::vector<std::pair<tree_node::node_id, tree_node::node_id>> deleted_edges = {};
+    greedyFAS(nodes, deleted_edges);
+    
+    ASAP(nodes);
+    
+    add_dummy_nodes(nodes, deleted_edges);
+    
+    coord_init(nodes);
+}
+
+const std::vector<tree_node::node_id> Net::get_sources() {
+    std::vector<tree_node::node_id> ids_sources = {};
+    deg_reinit(nodes);
+    for (int i = 0; i < nodes.size(); i++) {
+        if (nodes[i].deg_m == 0) {
+            ids_sources.push_back(nodes[i].id);
+        }
+    }
+    return ids_sources;
+}
+
+const std::vector<tree_node::node_id> Net::get_sinks() {
+    std::vector<tree_node::node_id> ids_sinks = {};
+    deg_reinit(nodes);
+    for (int i = 0; i < nodes.size(); i++) {
+        if (nodes[i].deg_p == 0) {
+            ids_sinks.push_back(nodes[i].id);
+        }
+    }
+    return ids_sinks;
+}
+
+const std::vector<tree_node::node_id> Net::get_successors(tree_node::node_id id) {
+    std::vector<tree_node::node_id> ids_successors = {};
+    for (int i = 0; i < nodes.size(); i++) {
+        if (nodes[i].id == id) {
+            ids_successors = nodes[i].succ;
+            break;
+        }
+    }
+    return ids_successors;
+}
+
+const std::vector<tree_node::node_id> Net::get_predecessors(tree_node::node_id id) {
+    std::vector<tree_node::node_id> ids_predecessors = {};
+    for (int i = 0; i < nodes.size(); i++) {
+        for (int j = 0; j < nodes[i].succ.size(); j++) {
+            if (nodes[i].succ[j] == id) {
+                ids_predecessors.push_back(nodes[i].id);
+            }
+        }
+    }
+    return ids_predecessors;
+}
```

Comment:
Actually we assume that `node_id` is an index into `Net::nodes` (add id validation into the mix):
```suggestion
const std::vector<tree_node::node_id> &Net::get_successors(tree_node::node_id id) {
   return nodes[id].succ;
}

const std::vector<tree_node::node_id> &Net::get_predecessors(tree_node::node_id id) {
    return nodes[id].pred;
}
```